app/src/components/ChainTrigger.tsx

```diff
@@ -0,0 +1,146 @@
+import useLookupName from '@/hooks/useLookupName'
+import { Chain } from '@/models/chain'
+import { ManualAddressInput } from '@/models/select'
+import { truncateAddress } from '@/utils/address'
+import { cn } from '@/utils/cn'
+import Image from 'next/image'
+import { ReactNode } from 'react'
+import { normalize } from 'viem/ens'
+import { useEnsAvatar } from 'wagmi'
+import CopyAddress from './ClipboardCopy'
+import ChainIcon from './svg/ChainIcon'
+import ChevronDown from './svg/ChevronDown'
+import Tooltip from './Tooltip'
+import VerticalDivider from './VerticalDivider'
+
+interface ChainTriggerProps {
+  value: Chain | null
+  onClick: () => void
+  error?: string
+  disabled?: boolean
+  className?: string
+  /** The action to display on the right side of the trigger. Like a wallet connect button. */
+  trailingAction?: ReactNode
+  /** The ref can be used to close the dropdown when the user clicks outside of it */
+  triggerRef?: React.RefObject<HTMLDivElement | null>
+  /** The input for manually entering an address. Used for specifying a recipient address. */
+  manualAddressInput?: ManualAddressInput
+  /** The connected account. */
+  walletAddress?: string
+}
+
+export default function ChainTrigger({
+  value,
+  error,
+  disabled,
+  onClick,
+  trailingAction,
+  className,
+  triggerRef,
+  manualAddressInput,
+  walletAddress,
+}: ChainTriggerProps) {
+  // wallet and ens
+  const addressLookup = useLookupName(value?.network, walletAddress?.toLowerCase())
+  const walletAddressShortened = walletAddress ? truncateAddress(walletAddress, 4, 4) : ''
+  const accountName = addressLookup ? addressLookup : walletAddressShortened
+  const { data: ensAvatar } = useEnsAvatar({
+    name: normalize(addressLookup || '') || undefined,
+  })
+
+  const shouldShowChainName =
+    (!walletAddress && (!manualAddressInput?.enabled || !manualAddressInput?.address)) ||
+    (manualAddressInput?.enabled && !manualAddressInput.address)
+
+  const shouldShowConnectedAccount = !manualAddressInput?.enabled && value
+
+  const handleManualAddressInputChange = (e: React.ChangeEvent<HTMLInputElement>) =>
+    manualAddressInput?.onChange?.({
+      enabled: manualAddressInput.enabled,
+      address: e.target.value,
+    })
+
+  const handleClick = () => {
```

Comment:
We could `useCallback` for event handlers to maintain referential equality between renders
```suggestion
  const handleClick = useCallback(() => {
    if (!disabled && onClick) onClick()
  }, [disabled, onClick])
```